From bd47f29847fcd6a68d359bc8aefb8749bb8a1b7c Mon Sep 17 00:00:00 2001
From: Thomas Gerbet <thomas.gerbet@enalean.com>
Date: Wed, 17 Nov 2021 15:57:19 +0100
Subject: [PATCH] request #24149: Indirect LDAP injection via the ldap_id
 attribute of a user

A taint annotation has been added to hightlight the issue. A more global
work on this topic will be done in independant contributions.

Change-Id: I969d0ba6d2ff85b418bcaf728f49afc78dd49571
---
 plugins/ldap/include/LDAP_DirectorySynchronization.class.php  | 2 +-
 src/common/DB/Compat/Legacy2018/LegacyDataAccessInterface.php | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/plugins/ldap/include/LDAP_DirectorySynchronization.class.php b/plugins/ldap/include/LDAP_DirectorySynchronization.class.php
index b9b6ea95c7d..a087f2a4e90 100755
--- a/plugins/ldap/include/LDAP_DirectorySynchronization.class.php
+++ b/plugins/ldap/include/LDAP_DirectorySynchronization.class.php
@@ -76,7 +76,7 @@ public function syncAll()
 
     public function ldapSync($row, $users_are_suspendable = true)
     {
-        $ldap_query = $this->ldap->getLDAPParam('eduid') . '=' . $row['ldap_id'];
+        $ldap_query = $this->ldap->getLDAPParam('eduid') . '=' . ldap_escape($row['ldap_id'], '', LDAP_ESCAPE_FILTER);
         $userSync   = $this->getLdapUserSync();
         $attributes = $userSync->getSyncAttributes($this->ldap);
 
diff --git a/src/common/DB/Compat/Legacy2018/LegacyDataAccessInterface.php b/src/common/DB/Compat/Legacy2018/LegacyDataAccessInterface.php
index c703bedf21d..00e584b8c77 100644
--- a/src/common/DB/Compat/Legacy2018/LegacyDataAccessInterface.php
+++ b/src/common/DB/Compat/Legacy2018/LegacyDataAccessInterface.php
@@ -177,6 +177,7 @@ public function fetch($result);
      * @param type $result
      *
      * @return type
+     * @psalm-taint-source ldap
      */
     public function fetchArray($result);
 
